feat: Embed SDK 2.0 Development (wip) (#210)
looker-open-source/embed-sdk: README.md

A typical setup might look like this. In this case, a dashboard with an id of `11` is created inside a DOM element with the id `embed_container`. The `dashboard:run:start` and `dashboard:run:complete` events are used to update the state of the embedding window's UI, and a button with an id of `run` is scripted to send a `dashboard:run` message to the dashboard.

```javascript
getEmbedSDK().init('looker.example.com', '/auth')

const setupConnection = (connection) => {
  document.querySelector('#run').addEventListener('click', () => {
    connection.asDashboardConnection().run()
  })
}

try {
  connection = await getEmbedSDK()

>     .createDashboardWithId(11)

Reviewer: this still wants numeric ids? or is it string | number?

    .appendTo('#embed_container')
    .on('dashboard:run:start', () => updateStatus('Running'))
    .on('dashboard:run:complete', () => updateStatus('Done'))
    .build()
    .connect()
  setupConnection(connection)
} catch (error) {
  console.error('An unexpected error occurred', error)
}
```

A more complete example can be found [here](https://github.com/looker-open-source/embed-sdk/demo/demo_single_frame.ts) and [here](https://github.com/looker-open-source/embed-sdk/demo/demo_multi_frame.ts).

## Details

The Looker Embed SDK uses the fluent interface pattern. The construction of the embedded content is broken into two phases, building and connecting. The hosting application may interact with the embedded content once the connection is established.

### Building

> Initialize the SDK with address of the Looker server and the endpoint of the embedding application server that will create a signed Looker embedded login URL (for private embedding, omit the signing endpoint).

Reviewer: ```suggestion
Initialize the SDK with the address of the Looker server and the endpoint of the embedding application server that will create a signed Looker embedded login URL (for private embedding, omit the signing endpoint).
```

```javascript
getEmbedSDK().init('looker.example.com', '/auth')
```

In this example, `/auth` is a backend service that must be implemented as described in the [Signed URL Auth Endpoint](#the-auth-endpoint) section.

After the SDK is initialized, create the builder using an `id` or `URL`. For example, to create a Dashboard embed builder do one of the following:

```javascript
getEmbedSDK().createDashboardWithId('42')
// OR
getEmbedSDK().createDashboardWithUrl('/embed/dashboards/42?state=california')
```

The `create[ContentType]WithUrl` functions can used with both signed and cookieless embed starting with Embed SDK 2.0.0. Prior to Embed SDK 2.0.0, Cookieless Embed only supported `create[CotentType]WithId`.

Additional attributes can be added to the builder to complete the setup:

```javascript
  .appendTo('#dashboard')
```

Event handlers can be added:

```javascript
  .on('dashboard:run:start',
      () => updateStatus('Running')
  )
  .on('dashboard:run:complete',
      () => updateStatus('Done')
  )
```

Create an embedded client by calling the build method:

```javascript
  .build()
```

### Connecting

> Once the client is built, call connect to create the IFRAME. The connect process creates the `src` attribute used for the actual IFRAME. How the `src` value is generated is based upon how the embed SDK is initialized.

Reviewer: Since I'm suggesting another edit:
```suggestion
Once the client is built, call `connect` to create the IFRAME. The connect process creates the `src` attribute used for the actual IFRAME. How the `src` value is generated is based upon how the embed SDK is initialized.
```

3. Private - the embed connection is private if the 2nd argument of the `init` call is not provided. In this case the URL is derived from the builder and decorated with the parameters required for Looker embed. For private embed, it is expected that the user is already logged into Looker OR that embedding URL include the `allow_login_screen=true` parameter.

`connect` returns a Promise that resolves to the connection interface for the embedded IFRAME.

```javascript
  .connect()
  .then((connection) => {
    // Save the connection
  })
  .catch(console.error)
```

### Interacting

Embed SDK 2.0.0 returns a unified connection that supports interacting with all Looker content types. The Embedding application can determine what kind of content is currently being displayed and interact accordingly.

```javascript
if (connection.getPageType() === 'dashboards') {
  connection.asDashboardConnection().run()
} else (connection.getPageType() === 'looks') {
  connection.asLookConnection().run()
} else (connection.getPageType() === 'explore') {

>   connection.asExploreConnection().run()

Reviewer: do you want to add reports and visualizations to the snippet for completeness?

Author: Reports do not support the run option as yet

```javascript
getEmbedSDK(new LookerEmbedExSDK()).init('looker.example.com', '/auth')
```

## Building URLs for the SDK

_The Embed SDK 2.0.0 now supports URL building for cookieless embed._

The main documentation for Looker SSO embed URLs is [here](https://docs.looker.com/r/sdk/sso-embed).

The Embed SDK will sanitize the URL if necessary:

1. It will strip hostname and protocol if present.
2. It will prefix the URL with `/embed` if missing.
3. It will add the `sdk` parameter if missing.
4. It will add the `embed_domain` parameter if missing.
5. It will throw an error if the URL is invalid.

`/embed/looks/4?embed_domain=https://mywebsite.com` becomes `/embed/looks/4?embed_domain=https://mywebsite.com&sdk=3`.

<a name="the-auth-endpoint" name"the-auth-endpoint"></a>

## Signed URL Auth Endpoint

This section does not apply to cookieless embed. See the [cookieless embed](#cookieless) section for details.

> In order to use the Embed SDK on the frontend you must supply a backend service that handles signing of the Embed URL. This service is called by the Embed SDK to generate a signed URL that is unique to the requesting user. The backend process can either generate the signed embed URL itself using an embed secret or can generate the URL by calling the [Looker Create Signed Embed URL API](https://cloud.google.com/looker/docs/reference/looker-api/latest/methods/Auth/create_sso_embed_url). Manual URL generation and signing avoids calling the Looker API resulting in decreased latency. Calling the Looker API requires less code and is easier to maintain.

Reviewer: ```suggestion
In order to use the Embed SDK on the frontend you must supply a backend service that handles signing of the Embed URL. This service is called by the Embed SDK to generate a signed URL that is unique to the requesting user. The backend process can either generate the signed embed URL itself using an embed secret or can generate the URL by calling the [Looker Create Signed Embed URL API](https://cloud.google.com/looker/docs/reference/looker-api/latest/methods/Auth/create_sso_embed_url). Manual URL generation and signing avoids calling the Looker API, resulting in decreased latency. Calling the Looker API requires less code and is easier to maintain.
```

Not a blocker

A typical setup might look like this. In this case, a dashboard with an id of `11` is created inside a DOM element with the id `dashboard`. The `dashboard:run:start` and `dashboard:run:complete` events are used to update the state of the embedding window's UI, and a button with an id of `run` is scripted to send a `dashboard:run` message to the dashboard.

```javascript
getEmbedSDK().init('looker.example.com', '/auth')

const setupDashboard = (connection) => {
  document.querySelector('#run').addEventListener('click', () => {
    connection.asDashboardConnection().run()
  })
}

try {
  connection = await getEmbedSDK()
    .createDashboardWithId(11)
    .appendTo('#embed_container')
    .on('dashboard:run:start', () => updateStatus('Running'))
    .on('dashboard:run:complete', () => updateStatus('Done'))
    .build()
    .connect()
  setupDashboard(connection)
} catch (error) {
  console.error('An unexpected error occurred', error)
}
```

> A more complete example can be found [here](demo/demo_single_frame.ts) and [here](demo/demo_multi_frame.ts). Detailed instructions on how to use the SDK can be found [here](#demo).

Reviewer: these links are not resolving on the npm site

Author: I've fixed this to use a full URL. It will be incorrect until this PR is merged

## Details

The Looker Embed SDK uses a fluent interface pattern. The construction of the embedded content is broken into two phases, building and connecting.

### Building

First initialize the SDK with address of the Looker server and the endpoint of the embedding application server that will create a signed Looker embedded login URL.

```javascript
getEmbedSDK().init('looker.example.com', '/auth')
```

> In this example, `/auth` is a backend service that must be implemented as described in the [Auth](#the-auth-endpoint) section.

Reviewer: this link isn't resolving

Author: Fixed.

3.

### Signed URL Frontend Process

This section does not apply to cookieless embed. See the [cookieless embed](#cookieless) section for details.

The _frontend_ process using the Embed SDK entails:

1. The embed SDK is initialized with the Looker host and the backend embed login signing endpoint:

```javascript
getEmbedSDK().init('looker.example.com', '/auth')
```

2. Build the embed definition

```javascript
const builder = getEmbedSDK()
  .createDashboardWithId(11)
  .append('#embed-container')
  .build()
```

3. Create the IFRAME

```javascript
const connection = await builder.connect()
```

The call to connect calls the auth endpoint to get a signed embed login URL. This URL is used to construct the IFRAME. The conection returned from the call allows the hosting application to interact with the Looker IFRAME. Embed SDK 2.0.x has a single connection object that can interact will Looker embed object types (dashboards, explores, looks, extensions). This means that it is no longer a requirement to replace an existing IFRAME with another IFRAME to display a different object type or object type instance.

> If another IFRAME is required, the SDK will not call the auth endpoint again as the embed session already exists. Instead the IFRAME will be created with the requested embed URL without signing a `login/embed` URL. If multiple IFRAMEs are created on page load (not recommended but it does work), the SDK will wait for the first IFRAME to be created before creating the more IFRAMEs. This is done to reduce contention that can occur when complex embed users are created.

Reviewer: ```suggestion
If another IFRAME is required, the SDK will not call the auth endpoint again as the embed session already exists. Instead the IFRAME will be created with the requested embed URL without signing a `login/embed` URL. If multiple IFRAMEs are created on page load (not recommended but it does work), the SDK will wait for the first IFRAME to be created before creating more IFRAMEs. This is done to reduce contention that can occur when complex embed users are created.
```

4. Load a different object type.

```javascript
await connection.loadLook('42')

await connection.loadUrl('/embed/dashboards/84?state=California')
```

> Different looker object types or object type instances can be loaded without recreating the IFRAME.

Reviewer: ```suggestion
Different Looker object types or object type instances can be loaded without recreating the IFRAME.
```

### Acquire session backend process

> This process is called every time a Looker Embed IFRAME is created. The acquire session backend process requires that the Looker api endpoint `acquire_embed_cookieless_session` be called to create an embed session or attach to an existing embed session. The endpoint accepts an embed user definition and creates or updates it. This is similar behavior to the signed URL embed login as they both can create and update embed user definitions.

Reviewer: ```suggestion
This process is called every time a Looker Embed IFRAME is created. The acquire session backend process requires that the Looker API endpoint `acquire_embed_cookieless_session` be called to create an embed session or attach to an existing embed session. The endpoint accepts an embed user definition and creates or updates it. This is similar behavior to the signed URL embed login as they both can create and update embed user definitions.
```

## Cookieless Embed

_Cookieless embed in Embed SDK 2.0.0 now supports the use of `withUrl`._

> Looker cookieless embed allows the Looker application to be embedded by an html page that is served from a different domain than the Looker host. With signed URL embed, to avoid third party cookie blocking, the Looker application must be served from a sub domain of the hosting application OR the user must enable third party cookies in the browser. Enabling cookieless embedding is documented in more detail [here](https://cloud.google.com/looker/docs/r/sdk/cookieless-embed). Cookieless embed is available with Looker version 22.20 and above.

Reviewer: link just goes to signed embedding, not any cookieless section

}
```

### Initializing the Embed SDK in the frontend

> Cookieless embed is initialized by calling `getEmbedSDK().initCookieless(...)` passing in the Looker host value and the the urls of the backend endpoints described previously. Once a Looker embed IFRAME is created it will communicate with the Embed SDK running in the host application and use the callbacks appropriately.

Reviewer: ```suggestion
Cookieless embed is initialized by calling `getEmbedSDK().initCookieless(...)` passing in the Looker host value and the urls of the backend endpoints described previously. Once a Looker embed IFRAME is created it will communicate with the Embed SDK running in the host application and use the callbacks appropriately.
```

### Embed domain allow list

Looker embed SSO requires that the host application domain be added to an allow list using the Looker configure admin page. Starting with Looker version `23.8`, the embed domain can be specified in the payload of the `acquire_embed_cookieless_session` call. This allows customers to dynamically specify the allowed domains. The allowed domain is associated with the embed session stored in the Looker server and is NOT persisted to the internal Looker database. This means that the embed domain must be included with every `acquire_embed_cookieless_session` call.

> When implementing it is HIGHLY recommended that that the implementor NOT trust any embed domain sent from the browser (in other words do not use the location.origin from the browser). Instead, the embed application should implement a mechanism in the server that maps a user to particular domain.

Reviewer: ```suggestion
When implementing it is HIGHLY recommended that the implementor NOT trust any embed domain sent from the browser (in other words do not use the location.origin from the browser). Instead, the embed application should implement a mechanism in the server that maps a user to particular domain.
```

## Demonstration Application

The `./demo` directory contains the frontend code for an embedded application. The following features can demonstrated:

1. Signed URL embedding using a single IFRAME. Multiple Looker object types and instances can be loaded using a single IFRAME.
2. Signed URL embedding using multiple IFRAMEs. An IFRAME can be created for a dashboard, look, explore and/extension.
3. Cookieless embedding using a single IFRAME. Multiple Looker object types and instances can be loaded using a single IFRAME.
4. Cookieless embedding using multiple IFRAMEs. An IFRAME can be created for a dashboard, look, explore and/extension.
5. Embed Message API embedding (this is provided for information purposes only and demonstrates how to embed Looker without using the Embed SDK).

### Step 1 - Enable Embedding in your Looker instance

Enabling signed URL embedding is documented in more detail [here](https://cloud.google.com/looker/docs/single-sign-on-embedding).
Enabling cookieless embedding is documented in more detail [here](https://cloud.google.com/looker/docs/r/sdk/cookieless-embed).

- Navigate to Admin > _Platform_ Embed on your Looker instance. This requires Admin privileges.

> - The demo server runs by default at [http://localhost:8080](http://localhost:8080). By adding that address to the "Embedded Domain Whitelist" you can enable the demo to receive messages from Looker.

Reviewer: ```suggestion
- The demo server runs by default at [http://localhost:8080](http://localhost:8080). By adding that address to the "Embedded Domain Allowlist" you can enable the demo to receive messages from Looker.
```

# Looker API server (include the protocol)
LOOKER_API_URL=https://mycompany.looker.com:19999
# Host name for the demo server
LOOKER_DEMO_HOST=localhost
# Port for the demo server
LOOKER_DEMO_PORT=8080
# Demo server cookie secret (used to encrypt the cookie)
COOKIE_SECRET=cookie_stash
# Verify the Looker certificate (for most embed developers, the Looker certificate will be valid)
LOOKER_VERIFY_SSL=true

# Looker Embed Configuration

# Embed type to demo
# signed - use signed embedding
# cookieless - use cookieless embedding
# private - use private embedding
LOOKER_EMBED_TYPE=signed
# Looker embed secret - from /admin/embed page
LOOKER_EMBED_SECRET=
# Looker client id - from /admin/users/api3_key/{id}
LOOKER_CLIENT_ID=
# Looker client secret - from /admin/users/api3_key/{id}
LOOKER_CLIENT_SECRET=
# Use dynamic embed domains for cookieless embed
LOOKER_USE_EMBED_DOMAIN=false

# Looker Embed Data Configuration
# Set to - if demo needs to ignore it

# Dashboard IDs
LOOKER_DASHBOARD_ID=1
LOOKER_DASHBOARD_ID_2=2
# Look ID
LOOKER_LOOK_ID=1
# Explore ID
LOOKER_EXPLORE_ID=thelook::orders
# Extension ID
LOOKER_EXTENSION_ID=extension::my-great-extension
# Report ID

> LOOKER_REPORT_ID=58a99d9e-eb3f-4c08-a0d9-4ac839a807b7

Reviewer: You want to keep this UUID in the doc?

# Query Visualization ID

> LOOKER_QUERY_VISUALIZATION_ID=O78fF73U3qNe3SelryY838

Reviewer: Keep this slug in the doc?

```

> - Edit the `demo/demo_user.json` file to be appropriate for the type of user you want to embed. Normally your backend service would use information about the user logged into your embedding application to inform Looker about important user properties that control data access controls. The `demo/demo_user.json` file is also used for cookieless embedding. Remember that cookieless_embed always treast `force_logout_login` as `true`. See [documentation](https://cloud.google.com/looker/docs/single-sign-on-embedding) for detailed information on the content of the embed user definition.

Reviewer: ```suggestion
- Edit the `demo/demo_user.json` file to be appropriate for the type of user you want to embed. Normally your backend service would use information about the user logged into your embedding application to inform Looker about important user properties that control data access controls. The `demo/demo_user.json` file is also used for cookieless embedding. Remember that cookieless_embed always treats `force_logout_login` as `true`. See [documentation](https://cloud.google.com/looker/docs/single-sign-on-embedding) for detailed information on the content of the embed user definition.
```

2. Signed URL embedding using multiple IFRAMES. An IFRAME can be created for a dashboard, look, explore and/extension.
3. Cookieless embedding using a single IFRAME. Multiple Looker object types and instances can be loaded using a single IFRAME.
4. Cookieless embedding using multiple IFRAMES. An IFRAME can be created for a dashboard, look, explore and/extension.
5. Embed Message API embedding (this is provided for information purposes only and demonstrates how to embed Looker without using the Embed SDK).

### Step 1 - Enable Embedding in your Looker instance

Enabling signed URL embedding is documented in more detail [here](https://cloud.google.com/looker/docs/single-sign-on-embedding).
Enabling cookieless embedding is documented in more detail [here](https://cloud.google.com/looker/docs/r/sdk/cookieless-embed).

- Navigate to Admin > _Platform_ Embed on your Looker instance. This requires Admin privileges.
- The demo server runs by default at [http://localhost:8080](http://localhost:8080). By adding that address to "Embedded Domain Whitelist" you can enabled the demo to receive messages from Looker.
- Turn on "Embed SSO Authentication"

> - In order to use embedding you must generate an "Embed Secret" for SSO embedding and/or a JWT secret for cookieless embedding. A Looker instance can support signed URL and cookieles embedding simulaneously.

Reviewer: ```suggestion
- In order to use embedding you must generate an "Embed Secret" for SSO embedding and/or a JWT secret for cookieless embedding. A Looker instance can support signed URL and cookieless embedding simultaneously.
```

also, the create_sso_embed_url endpoint will now:
- enable signed embedding if it's off
- add the embed domain(s) if they're missing
- create an Embed secret if required
- disable same origin if it's on

localStorage.debug = 'looker:chatty:*'
```

> The hosting window and the embedded IFRAME have separate local storage, so you can enable logging on one, the other or both. You can disable logging with

Reviewer: ```suggestion
The hosting window and the embedded IFRAME have separate local storage, so you can enable logging on one, the other, or both. You can disable logging with
```
